### PR TITLE
Add gene search component to Species Selector

### PR DESCRIPTION
**Pull request by azangru**
## Description
This PR includes a first take at adding a modal gene search element. I created it before I realised how much this modal element is reused in other latest designs. As a result, the PR contains one-off solutions that will subsequently have to be refactored (e.g. we will need some shared component for the grey panel; we will probably have to keep the state of whether the panel is opened in redux, etc.).

**The new functionality added in this PR:**
- "Find a gene" button added to the end of the species lozenges
- When clicked, it opens a gene search modal view
- Gene is searched against all currently selected species

**Missing** _(haven't been included in designs):_
- Loading state
- Error state

## Related JIRA Issue(s)
https://www.ebi.ac.uk/panda/jira/browse/ENSWBSITES-1915

## Deployment URL(s)
http://species-selector-find-gene.review.ensembl.org

### CODE REVIEW

**Review by ens-st3 — APPROVED** [review]
Observations left in JIRA

File: src/shared/components/gene-search-panel/GeneSearchPanel.scss [hunk 1]
```diff
@@ -0,0 +1,111 @@
+@import 'src/styles/common';
+
+.geneSearchPanel {
+  height: 100%;
+  background-color: $light-grey;
+  // TODO: Can we find a better name for $global-padding-left? It is not very global, and it is pretty large
+  padding: 65px 0 $global-padding-bottom $global-padding-left; // TODO: the 65px on top looks like it is something repeatable between components?
```
Comment (imransl) [inline src/shared/components/gene-search-panel/GeneSearchPanel.scss:7]: `common-padding-left`?

File: src/shared/components/gene-search-panel/GeneSearchPanel.scss [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+.resultsTable th {
+  text-align: left;
+  border-bottom: 1px solid $blue;
+  padding: 0 0.5rem 0.4rem 1rem;
+  font-weight: $light;
+}
+
+.resultsTable td {
+  padding: 0.4rem 0.5rem 0.4rem 1rem;
```
Comment (imransl) [inline src/shared/components/gene-search-panel/GeneSearchPanel.scss:50]: Aren't some of these values used elsewhere for padding and margin? Should these be common values?

File: src/shared/components/gene-search-panel/GeneSearchPanel.tsx [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+        />
+        <span className={radioStyles.label}>Only species in list</span>
+      </div>
+      <div
+        className={classNames(
+          radioStyles.radioGroupItem,
+          styles.pseudoRadioDisabled
+        )}
+      >
+        <span className={classNames(radioStyles.radio)} />
```
Comment (imransl) [inline src/shared/components/gene-search-panel/GeneSearchPanel.tsx:130]: There's just one class used here so no need of `classNames`?

**Comment by jyothishnt** [comment]
Scroll bar needs to be applied on to the results container.

File: src/shared/components/gene-search-panel/GeneSearchPanel.tsx [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+            <span className={styles.speciesCount}>
+              {groupedSearchMatches.length}
+            </span>
+            species
+          </th>
+          <th>Gene</th>
+        </tr>
+      </thead>
+      <tbody>
+        {groupedSearchMatches.map((rowData) => (
```
Comment (jyothishnt) [inline src/shared/components/gene-search-panel/GeneSearchPanel.tsx:161]: `rowData` -> `match`?

File: src/shared/components/gene-search-panel/GeneSearchPanel.tsx [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+
+  const urlForEntityViewer = urlFor.entityViewer({
+    genomeId: genomeIdForUrl,
+    entityId: buildFocusIdForUrl({
+      type: 'gene',
+      objectId: unversioned_stable_id
+    })
+  });
+
+  // NOTE: If this panel opens in genome browser, we should add `replaceState: true`
```
Comment (jyothishnt) [inline src/shared/components/gene-search-panel/GeneSearchPanel.tsx:236]: What is the use of this?

File: src/shared/components/gene-search-panel/GeneSearchPanel.tsx [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+    searchMatchesMap.set(species.genome_id, {
+      speciesInfo: species,
+      searchMatches: []
+    });
+  });
+
+  searchMatches.forEach((match) => {
+    const matches = searchMatchesMap.get(match.genome_id)?.searchMatches;
+    if (matches) {
+      matches.push(match);
```
Comment (jyothishnt) [inline src/shared/components/gene-search-panel/GeneSearchPanel.tsx:286]: Where is `matches` used?

File: src/content/app/species-selector/SpeciesSelector.scss [hunk 7]
```diff
@@ -0,0 +1,6 @@
+.grid {
+  display: grid;
+  grid-template-rows: max-content 1fr;
+  height: 100%;
+  overflow: auto; // FIXME: scrolling should start happen below the app bar
```
[the remark this answers is not in the document]
Comment (azangru) [reply src/content/app/species-selector/SpeciesSelector.scss:5]: Yes. This shouldn't happen; but I don't know how we are going to treat this

https://user-images.githubusercontent.com/6834224/234598821-4761deec-58a1-4d83-807b-52ed9dbf2d47.mov

File: src/content/app/species-selector/components/species-selector-app-bar/SpeciesSelectorAppBar.tsx [hunk 8]
```diff
@@ -68,15 +77,30 @@ const SelectedSpeciesList = (props: { selectedSpecies: CommittedItem[] }) => {
...
   ));
 
-  return <SpeciesTabsWrapper speciesTabs={selectedSpecies} />;
+  const geneSearchButton = props.isGeneSearchMode ? (
+    <GeneSearchCloseButton
+      key="find-a-gene"
+      onClick={props.onGeneSearchToggle}
+    />
+  ) : (
+    <GeneSearchButton key="find-a-gene" onClick={props.onGeneSearchToggle} />
```
[the remark this answers is not in the document]
Comment (azangru) [reply src/content/app/species-selector/components/species-selector-app-bar/SpeciesSelectorAppBar.tsx:98]: This element now becomes a member of an array of other elements (species lozenges). This is temporary; until we change the selected species component in the top bar. But in the meantime, for a keyless element in an array of elements, React would complain:

image

File: src/shared/components/gene-search-panel/GeneSearchPanel.tsx (hunk 5 above)
Comment (azangru) [reply src/shared/components/gene-search-panel/GeneSearchPanel.tsx:236]: Whether a new history item gets pushed to the browser history stack, or whether it replaces the topmost history item on the stack:

image

What it means in practical terms is that it determines what happens when you press on web browser's Back button. If you've added the new location via push, you go to location 2 on the diagram above; if via replace, then to location 1.

On the genome browser page, in-app search is using the `replace` mechanism. Maybe we don't need it (we don't use replace when we switch from gene to gene via zmenu or via previously viewed links)

File: src/shared/components/gene-search-panel/GeneSearchPanel.tsx (hunk 6 above)
Comment (azangru) [reply src/shared/components/gene-search-panel/GeneSearchPanel.tsx:286]: Good point. I can't remember what this was for. Perhaps it's remnant from before I used speciesList.forEach above. I'll remove it.

Comment (azangru) [reply src/shared/components/gene-search-panel/GeneSearchPanel.tsx:286]: `matches` is an array that's seating in searchMatchesMap (line 279 above) 😃 On line 283 we iterate through all search matches (a flat array of search matches, not grouped by species), then, for each match, on line 284, we look up in the map (filled on lines 277-279) an entry under the same genome id that a search match has; and then we push the search match into that array. As a result we distribute the flat array of search matches into groups, each keyed by the genome id.

File: src/shared/components/gene-search-button/GeneSearchButton.tsx [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React from 'react';
+
+import SearchIcon from 'static/icons/icon_search.svg';
```
[the remark this answers is not in the document]
Comment (azangru) [reply src/shared/components/gene-search-button/GeneSearchButton.tsx:19]: Updated the styles.

File: src/shared/components/gene-search-button/GeneSearchCloseButton.tsx [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+import styles from './GeneSearchButton.scss';
+
+type Props = {
+  onClick: () => void;
+};
+
+const GeneSearchButton = (props: Props) => {
+  return (
+    <button onClick={props.onClick} className={styles.findGeneButton}>
+      <CloseIcon />
```
[the remark this answers is not in the document]
Comment (azangru) [reply src/shared/components/gene-search-button/GeneSearchCloseButton.tsx]: Updated the styles.

File: src/shared/components/gene-search-panel/GeneSearchPanel.tsx (hunk 4 above)
Comment (azangru) [reply src/shared/components/gene-search-panel/GeneSearchPanel.tsx]: ✅

**Comment by azangru** [comment]
> Scroll bar needs to be applied on to the results container.

I've updated the layout to make the table of results scrollable.

File: src/content/app/species-selector/SpeciesSelector.scss (hunk 7 above)
[the remark this answers is not in the document]
Comment (jyothishnt) [reply src/content/app/species-selector/SpeciesSelector.scss:5]: Did you mean scrolling should happen below the species bar?

[the remark this answers is not in the document]
Comment (azangru) [reply src/content/app/species-selector/SpeciesSelector.scss:5]: Yes.

File: src/shared/components/gene-search-panel/GeneSearchPanel.tsx (hunk 6 above)
Comment (jyothishnt) [reply src/shared/components/gene-search-panel/GeneSearchPanel.tsx:312]: OK, so `matches` is a reference?

File: src/content/app/species-selector/SpeciesSelector.scss (hunk 7 above)
Comment (jyothishnt) [inline src/content/app/species-selector/SpeciesSelector.scss:5]:
```suggestion
  overflow: auto; // FIXME: scrolling should start happen below the species bar
```

Comment (azangru) [reply src/content/app/species-selector/SpeciesSelector.scss]: ✅

File: src/shared/components/gene-search-panel/GeneSearchPanel.tsx (hunk 6 above)
Comment (azangru) [reply src/shared/components/gene-search-panel/GeneSearchPanel.tsx:312]: Yes, line numbers are now off. This is what the matches are

image

**Review by jyothishnt — APPROVED** [review]

**Pull request merged**